source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -165,26 +172,86 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   return Network::PostIoAction::KeepOpen;
 }
 
-Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
-  Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
-  if (!end_stream_read_ && !read_error_) {
-    result = raw_buffer_socket_->doRead(raw_read_buffer_);
-    ENVOY_CONN_LOG(debug, "TSI: raw read result action {} bytes {} end_stream {}",
-                   callbacks_->connection(), enumToInt(result.action_), result.bytes_processed_,
-                   result.end_stream_read_);
-    if (result.action_ == Network::PostIoAction::Close && result.bytes_processed_ == 0) {
-      return result;
+Network::IoResult TsiSocket::repeatReadAndUnprotect(Buffer::Instance& buffer,
+                                                    Network::IoResult prev_result) {
+  Network::IoResult result = prev_result;
+  uint64_t total_bytes_processed = 0;
+
+  while (true) {
+    // Do unprotect.
+    if (raw_read_buffer_.length() > 0) {
+      uint64_t prev_size = buffer.length();
+      ENVOY_CONN_LOG(debug, "TSI: unprotecting buffer size: {}", callbacks_->connection(),
+                     raw_read_buffer_.length());
+      tsi_result status = frame_protector_->unprotect(raw_read_buffer_, buffer);
+      if (status != TSI_OK) {
+        ENVOY_CONN_LOG(debug, "TSI: unprotect failed: status: {}", callbacks_->connection(),
+                       status);
+        result.action_ = Network::PostIoAction::Close;
+        break;
```

Comment:
No test coverage for this branch, could you add tests?

https://storage.googleapis.com/envoy-pr/0357f2f/coverage/source/extensions/transport_sockets/alts/tsi_socket.cc.gcov.html

Reply:
Done.